// @carbon/react/es/components/Checkbox/Checkbox.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { objectWithoutProperties as _objectWithoutProperties, defineProperty as _defineProperty, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import PropTypes from 'prop-types';
import React__default from 'react';
import cx from 'classnames';
import '../Text/index.js';
import { usePrefix } from '../../internal/usePrefix.js';
import { WarningFilled, WarningAltFilled } from '@carbon/icons-react';
import setupGetInstanceId from '../../tools/setupGetInstanceId.js';
import { Text } from '../Text/Text.js';

var _excluded = ["className", "helperText", "id", "labelText", "onChange", "onClick", "indeterminate", "invalid", "invalidText", "hideLabel", "readOnly", "title", "warn", "warnText"];
var getInstanceId = setupGetInstanceId();
var Checkbox = /*#__PURE__*/React__default.forwardRef(function (_ref, _ref2) {
  var _classNames;

  var className = _ref.className,
      helperText = _ref.helperText,
      id = _ref.id,
      labelText = _ref.labelText,
      _onChange = _ref.onChange,
      _onClick = _ref.onClick,
      indeterminate = _ref.indeterminate,
      invalid = _ref.invalid,
      invalidText = _ref.invalidText,
      hideLabel = _ref.hideLabel,
      readOnly = _ref.readOnly,
      _ref$title = _ref.title,
      title = _ref$title === void 0 ? '' : _ref$title,
      warn = _ref.warn,
      warnText = _ref.warnText,
      other = _objectWithoutProperties(_ref, _excluded);

  var prefix = usePrefix();
  var showWarning = !readOnly && !invalid && warn;
  var showHelper = !invalid && !warn;
  var checkboxGroupInstanceId = getInstanceId();
  var helperId = !helperText ? undefined : "checkbox-helper-text-".concat(checkboxGroupInstanceId);
  var helper = helperText ? /*#__PURE__*/React__default.createElement("div", {
    id: helperId,
    className: "".concat(prefix, "--form__helper-text")
  }, helperText) : null;
  var wrapperClasses = cx("".concat(prefix, "--form-item"), "".concat(prefix, "--checkbox-wrapper"), className, (_classNames = {}, _defineProperty(_classNames, "".concat(prefix, "--checkbox-wrapper--readonly"), readOnly), _defineProperty(_classNames, "".concat(prefix, "--checkbox-wrapper--invalid"), !readOnly && invalid), _defineProperty(_classNames, "".concat(prefix, "--checkbox-wrapper--warning"), showWarning), _classNames));
  var innerLabelClasses = cx("".concat(prefix, "--checkbox-label-text"), _defineProperty({}, "".concat(prefix, "--visually-hidden"), hideLabel));
  return /*#__PURE__*/React__default.createElement("div", {
    className: wrapperClasses
  }, /*#__PURE__*/React__default.createElement("input", _extends({}, other, {
    type: "checkbox",
    "data-invalid": invalid ? true : undefined,
    onChange: function onChange(evt) {
      if (!readOnly && _onChange) {
        _onChange(evt, {
          checked: evt.target.checked,
          id: id
        });
      }
    },
    className: "".concat(prefix, "--checkbox"),
    id: id,
    ref: function ref(el) {
      if (el && indeterminate) {
        el.indeterminate = indeterminate;
      }

      if (typeof _ref2 === 'function') {
        _ref2(el);
      } else if (_ref2 && Object(_ref2) === _ref2) {
        _ref2.current = el;
      }
    } // readonly attribute not applicable to type="checkbox"
    // see - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox
    ,
    "aria-readonly": readOnly,
    onClick: function onClick(evt) {
      if (readOnly) {
        // prevent default stops the checkbox being updated
        evt.preventDefault();
      } // pass onClick event on to the user even if readonly


      if (_onClick) {
        _onClick(evt);
      }
    }
  })), /*#__PURE__*/React__default.createElement("label", {
    htmlFor: id,
    className: "".concat(prefix, "--checkbox-label"),
    title: title
  }, /*#__PURE__*/React__default.createElement(Text, {
    className: innerLabelClasses
  }, labelText)), /*#__PURE__*/React__default.createElement("div", {
    className: "".concat(prefix, "--checkbox__validation-msg")
  }, !readOnly && invalid && /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(WarningFilled, {
    className: "".concat(prefix, "--checkbox__invalid-icon")
  }), /*#__PURE__*/React__default.createElement("div", {
    className: "".concat(prefix, "--form-requirement")
  }, invalidText)), showWarning && /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(WarningAltFilled, {
    className: "".concat(prefix, "--checkbox__invalid-icon ").concat(prefix, "--checkbox__invalid-icon--warning")
  }), /*#__PURE__*/React__default.createElement("div", {
    className: "".concat(prefix, "--form-requirement")
  }, warnText))), showHelper && helper);
});
Checkbox.propTypes = {
  /**
   * Specify whether the underlying input should be checked
   */
  checked: PropTypes.bool,

  /**
   * Specify an optional className to be applied to the <label> node
   */
  className: PropTypes.string,

  /**
   * Specify whether the underlying input should be checked by default
   */
  defaultChecked: PropTypes.bool,

  /**
   * Specify whether the Checkbox should be disabled
   */
  disabled: PropTypes.bool,

  /**
   * Provide text for the form group for additional help
   */
  helperText: PropTypes.node,

  /**
   * Specify whether the label should be hidden, or not
   */
  hideLabel: PropTypes.bool,

  /**
   * Provide an `id` to uniquely identify the Checkbox input
   */
  id: PropTypes.string.isRequired,

  /**
   * Specify whether the Checkbox is in an indeterminate state
   */
  indeterminate: PropTypes.bool,

  /**
   * Specify whether the Checkbox is currently invalid
   */
  invalid: PropTypes.bool,

  /**
   * Provide the text that is displayed when the Checkbox is in an invalid state
   */
  invalidText: PropTypes.node,

  /**
   * Provide a label to provide a description of the Checkbox input that you are
   * exposing to the user
   */
  labelText: PropTypes.node.isRequired,

  /**
   * Provide an optional handler that is called when the internal state of
   * Checkbox changes. This handler is called with event and state info.
   * `(event, { checked, id }) => void`
   */
  onChange: PropTypes.func,

  /**
   * Specify whether the Checkbox is read-only
   */
  readOnly: PropTypes.bool,

  /**
   * Specify a title for the <label> node for the Checkbox
   */
  title: PropTypes.string,

  /**
   * Specify whether the Checkbox is currently in warning state
   */
  warn: PropTypes.bool,

  /**
   * Provide the text that is displayed when the Checkbox is in warning state
   */
  warnText: PropTypes.node
};
Checkbox.defaultProps = {
  onChange: function onChange() {},
  indeterminate: false
};
Checkbox.displayName = 'Checkbox';

export { Checkbox as default };
