// @carbon/react/es/components/InlineCheckbox/InlineCheckbox.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { defineProperty as _defineProperty } from '../../_virtual/_rollupPluginBabelHelpers.js';
import PropTypes from 'prop-types';
import React__default, { useRef, useEffect } from 'react';
import deprecate from '../../prop-types/deprecate.js';
import { usePrefix } from '../../internal/usePrefix.js';
import { useMergedRefs } from '../../internal/useMergedRefs.js';

var _InlineCheckbox$propT;
/** @type any */

var InlineCheckbox = /*#__PURE__*/React__default.forwardRef(function InlineCheckbox(props, forwardRef) {
  var ariaLabel = props['aria-label'],
      deprecatedAriaLabel = props.ariaLabel,
      _props$checked = props.checked,
      checked = _props$checked === void 0 ? false : _props$checked,
      disabled = props.disabled,
      id = props.id,
      indeterminate = props.indeterminate,
      name = props.name,
      _props$onChange = props.onChange,
      _onChange = _props$onChange === void 0 ? function () {} : _props$onChange,
      onClick = props.onClick,
      onKeyDown = props.onKeyDown,
      title = props.title;

  var prefix = usePrefix();
  var inputRef = useRef(null);
  var ref = useMergedRefs([inputRef, forwardRef]);
  var inputProps = {
    checked: checked,
    className: "".concat(prefix, "--checkbox"),
    disabled: disabled,
    id: id,
    name: name,
    onClick: onClick,
    onChange: function onChange(evt) {
      _onChange(evt.target.checked, id, evt);
    },
    onKeyDown: onKeyDown,
    ref: ref,
    type: 'checkbox'
  };

  if (indeterminate) {
    inputProps.checked = false;
  }

  useEffect(function () {
    inputRef.current.indeterminate = indeterminate;
  }, [indeterminate]);
  return /*#__PURE__*/React__default.createElement("div", {
    className: "".concat(prefix, "--checkbox--inline")
  }, /*#__PURE__*/React__default.createElement("input", inputProps),
  /*#__PURE__*/

  /* eslint-disable jsx-a11y/label-has-for,jsx-a11y/label-has-associated-control,jsx-a11y/click-events-have-key-events,jsx-a11y/no-noninteractive-element-interactions */
  React__default.createElement("label", {
    htmlFor: id,
    className: "".concat(prefix, "--checkbox-label"),
    "aria-label": deprecatedAriaLabel || ariaLabel,
    title: title,
    onClick: function onClick(evt) {
      evt.stopPropagation();
    }
  }));
});
InlineCheckbox.propTypes = (_InlineCheckbox$propT = {}, _defineProperty(_InlineCheckbox$propT, 'aria-label', PropTypes.string.isRequired), _defineProperty(_InlineCheckbox$propT, "ariaLabel", deprecate(PropTypes.string.isRequired, 'The `ariaLabel` prop has been deprecated in favor of `aria-label`. This prop will be removed in the next major release.')), _defineProperty(_InlineCheckbox$propT, "checked", PropTypes.bool), _defineProperty(_InlineCheckbox$propT, "disabled", PropTypes.bool), _defineProperty(_InlineCheckbox$propT, "id", PropTypes.string.isRequired), _defineProperty(_InlineCheckbox$propT, "indeterminate", PropTypes.bool), _defineProperty(_InlineCheckbox$propT, "name", PropTypes.string.isRequired), _defineProperty(_InlineCheckbox$propT, "onChange", PropTypes.func), _defineProperty(_InlineCheckbox$propT, "onClick", PropTypes.func), _defineProperty(_InlineCheckbox$propT, "onKeyDown", PropTypes.func), _defineProperty(_InlineCheckbox$propT, "title", PropTypes.string), _InlineCheckbox$propT);
var InlineCheckbox$1 = InlineCheckbox;

export { InlineCheckbox$1 as default };
